# Output description for rnaseq variant calling pipeline
This report described the process to call variants from RNA-seq processed data.

## Pipeline overview
The pipeline follows these steps:

* <a href="#picard">PICARD</a> v.2.18.27 - AddOrReplaceGroups and RemoveDuplicates
* <a href="#gatk">GATK</a> v.3.8.0 - Variant calling
* <a href="#snpeff">SnpEff</a> v.4.3k - Variant annotation
* <a href="#bcftools">Bcftools</a> v.1.9 - Merge VCF files
* <a href="#SnpSift">SnpSift</a> v.4.3k - Extract annotation fields.

**Note** All the results from the Variant Calling process are located in the corresponding `11-variant_calling` folder.

## Picard
[Picard](https://broadinstitute.github.io/picard/index.html)<a name="picard"> </a><a href="#picard_reference">[6]</a> is a set of command line tools for manipulating high-throughput sequencing (HTS) data. In this case we used the following Picard modules:
- AddOrReplaceReadGroups: Assigns all the reads in a file to a single new read-group.
- MarkDuplicates: Remove duplicated reads

**Output directory:** `01-markDuplicates/`

* `<sample_name>/<sample_name>_sorted.bam`: BAM file comming from AddOrReplaceGroups step.
* `<sample_name>/<sample_name>_output.metrics`: Metrics/statistics file comming from removing duplicated reads.
* `<sample_name>/<sample_name>_dedup.bam`: BAM file with duplicated reads removed.
* `<sample_name>/<sample_name>_dedup.bam.bai`: Index of the BAM file.

**Note** .bam files are veru big so by default are not sent to the scientist. If you need them, ask for them.

Picard documentation: [Picard docs](https://broadinstitute.github.io/picard/command-line-overview.html)

## GATK
[GATK](https://gatk.broadinstitute.org/hc/en-us) is a Genome Analysis Toolkit for ariant Discovery in High-Throughput Sequencing Data. GATK has developped a [best practice pipeline for variant calling in RNA-seq data](https://gatk.broadinstitute.org/hc/en-us/articles/360035531192-RNAseq-short-variant-discovery-SNPs-Indels-) which we followed. The modules used for this aim are:
1. SplitNCigarReads: Because RNA aligners have different conventions than DNA aligners, we need to reformat some of the alignments that span introns for HaplotypeCaller. This step splits reads with N in the cigar into multiple supplementary alignments and hard clips mismatching overhangs. By default this step also reassigns mapping qualities for good alignments to match DNA conventions.
2. RealignerTargetCreator: takes a coordinate-sorted and indexed BAM and a VCF of known indels and creates a target intervals file.
3. IndelRealigner: takes a coordinate-sorted and indexed BAM and a target intervals file generated by RealignerTargetCreator.
4. BaseRecalibrator: Generates recalibration table for Base Quality Score Recalibration (BQSR)
5. PrintReads: Print reads in the SAM/BAM/CRAM file.
6. HaplotypeCaller: Call SNPs and indels via local re-assembly of haplotypes
7. VariantFiltration: Filter variant calls based on INFO and/or FORMAT annotations

**Output directory:** `02-variants/`

* `<sample_name>/<sample_name>_split.bam`: BAM file comming from SplitNCigarReads step.
* `<sample_name>/<sample_name>_forIndelRealigner.intervals`: Intervals file created by RealignerTargetCreator:.
* `<sample_name>/<sample_name>_realigned.bam`: Realigned BAM file from IndelRealigner step.
* `<sample_name>/<sample_name>_recal_data.table`: Reclaibration table created by BaseRecalibrator.
* `<sample_name>/<sample_name>_recalibrated.bam`: Recalibrated BAM file from PrintReads step.
* `<sample_name>/<sample_name>.vcf`: VCF file from HaplotypeCaller variant calling.
* `<sample_name>/<sample_name>_filtered.vcf`: Filtered VCF file from VariantFiltration step.

**Note** .bam files are veru big so by default are not sent to the scientist. If you need them, ask for them.

## SnpEff

[SnpEff](http://snpeff.sourceforge.net/SnpEff.html) is a genetic variant annotation and functional effect prediction toolbox. It annotates and predicts the effects of genetic variants on genes and proteins (such as amino acid changes).

**Output directory:** `02-variants/`

* `<sample_name>/<sample_name>.ann.vcf`: Annotated VCF file.
* `<sample_name>/<sample_name>.ann.vcf.gz`: Annotated compressed VCF file.
* `<sample_name>/<sample_name>.ann.vcf.gz.idx`: Annotated compressed VCF file index.

## Bcftools

[BCFtools](http://samtools.github.io/bcftools/bcftools.html) can be used for different purposes. In this case we are using it to merge all the .ann.vcf.gz files into one.

**Output directory:** `02-variants/`

* `merged_annot_variants.vcf`: VCF file with the annotation of variants for all the samples.

## SnpSift

[SnpSift](http://snpeff.sourceforge.net/SnpSift.html) annotates genomic variants using databases, filters, and manipulates genomic annotated variants. After annotation with SnpEff, you can use SnpSift to help filter large genomic datasets in order to find the most significant variants.

**Output directory:** `02-variants/`

* `merged_annot_variants.table`: VCF file in table with the annotation of variants for all the samples, with the selected fields.

## Post-Processing
The resulting `merged_annot_variants.table` is furtherly processed in R to rename the file headers and filter variants based in the GATK score ("PASS") and the EFFECT of tje varints ("HIGH & MODERATE").

**Output directory:** `03-filtering/`

* `merged_annot_variants_reheader.table`:  Same as merged_annot_variants.table but with the header names fixed.
* `filtered_variants.txt`: Same as merged_annot_variants_reheader.table but filtered the variants by "PASS", and EFFECT "HIGH" & "MODERATE"
